agent/app/agent.go

```diff
@@ -454,11 +462,21 @@ func (agent *ecsAgent) setClusterInConfig(previousCluster string) error {
 
 // getEC2InstanceID gets the EC2 instance ID from the metadata service
 func (agent *ecsAgent) getEC2InstanceID() string {
-	instanceID, err := agent.ec2MetadataClient.InstanceID()
+	var instanceID string
+	var err error
+	backoff := retry.NewExponentialBackoff(getInstanceIdBackoffMin, getInstanceIdBackoffMax, getInstanceIdBackoffJitter, getInstanceIdBackoffMultiple)
+	for i := 0; i < 3; i++ {
+		instanceID, err = agent.ec2MetadataClient.InstanceID()
```

Comment:
do we know the error types returned by this API? are they classified as retriable/non retriable?

Reply:
Observed `404 - Not Found` HTTP error code from the logs
```
level=error time=2021-03-04T00:00:08Z msg="Unable to retrieve Host Instance PublicIPv4 Address: EC2MetadataError: failed to make EC2Metadata request\n\tstatus code: 404, request id: \ncaused by: <?xml version=\"1.0\" encoding=\"iso-8859-1\"?>\n<!DOCTYPE html PUBLIC \"-//W3C//DTD XHTML 1.0 Transitional//EN\"\n\t\"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd\">\n<html xmlns=\"http://www.w3.org/1999/xhtml\" xml:lang=\"en\" lang=\"en\">\n <head>\n  <title>404 - Not Found</title>\n </head>\n <body>\n  <h1>404 - Not Found</h1>\n </body>\n</html>\n" module=agent.go
```
And these are the errors can be returned by IMDSv2 API
```
400 - Missing or Invalid Parameters – The PUT request is not valid.

401 - Unauthorized – The GET request uses an invalid token. The recommended action is to generate a new token.

403 - Forbidden – The request is not allowed or the instance metadata service is turned off.

404 - Not Found – If the resource is not available.

Query throttling - when it exceeds rate limiting
```

Do we want to only retry on Query throttling and 404 error?

Reply:
its ok to keep this as is for now, since number of retries is less too.